### PR TITLE
[expo run:android] add flavors support

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Appears to resolve all `expo run:android` issues, it's unclear why we didn't make it like this originally.

- https://github.com/expo/expo-cli/issues/3846
- https://github.com/expo/expo-cli/issues/3845
- https://github.com/expo/expo-cli/issues/3714

# How

- Use forward look ahead regex split to determine which flavors are added to the build type. Use the discovered flavors to discover the correct binary output location.
- Fallback onto gradle install when we can't find the binary
- Fix opening app on device -- will split out

# Test Plan

- TBD

### CODE REVIEW

File: packages/expo-cli/src/commands/run/android/runAndroid.ts [hunk 1]
```diff
@@ -101,9 +102,16 @@ async function resolveOptionsAsync(
...
-  const apkDirectory = Android.getAPKDirectory(projectRoot);
-  const apkVariantDirectory = path.join(apkDirectory, variant);
+  // TODO: why would this be different? Can we get the different name?
+  const appName = 'app';
+
+  const apkDirectory = path.join(projectRoot, 'android', appName, 'build', 'outputs', 'apk');
+
+  // buildDeveloperTrust -> build, developer, trust (where developer, and trust are flavors).
+  const [buildType, ...flavorDimensions] = options.variant
+    .split(/(?=[A-Z])/)
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/run/android/runAndroid.ts:112]: could use some comment

File: packages/expo-cli/src/commands/run/android/runAndroid.ts [hunk 2]
```diff
@@ -101,9 +102,16 @@ async function resolveOptionsAsync(
...
-  const variant = options.variant.toLowerCase();
-  const apkDirectory = Android.getAPKDirectory(projectRoot);
-  const apkVariantDirectory = path.join(apkDirectory, variant);
+  // TODO: why would this be different? Can we get the different name?
+  const appName = 'app';
+
+  const apkDirectory = path.join(projectRoot, 'android', appName, 'build', 'outputs', 'apk');
+
+  // buildDeveloperTrust -> build, developer, trust (where developer, and trust are flavors).
+  const [buildType, ...flavorDimensions] = options.variant
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/run/android/runAndroid.ts:111]: Naming here is partially incorrect, if you have command `:app:buildFooBarRelease`
- app is an android module
- `foo` and `bar` are flavors (not flavorDimensions), but actual flavours can be named `Foo` and `Bar` we don't know that just based on the gradle command
-  flavorDimensions defines order of the flavors in the command, so if you have in your build.gradle `flavourDimensions: "dim1" "dim2"` then `foo` has to have dimension set to `dim1` and `bar` to `dim2`
- variant in this example would be `FooBarRelease`, build/bundle/assemble... is not the part of the variant
- buildType is either debug, release or some custom value

File: packages/expo-cli/src/commands/run/android/runAndroid.ts (hunk 1 above)
Comment (wkozyra95) [reply packages/expo-cli/src/commands/run/android/runAndroid.ts:112]: also, I'm not saying we should care about that case, but just FYI

if you have flavor e.g. `exaMPle` the gradle command would look like this `:app:buildExaMPleRelease`

Comment (EvanBacon) [reply packages/expo-cli/src/commands/run/android/runAndroid.ts:112]: I think for the time being, the current implementation is a fair compromise.

File: packages/expo-cli/src/commands/run/android/runAndroid.ts (hunk 2 above)
Comment (EvanBacon) [reply packages/expo-cli/src/commands/run/android/runAndroid.ts]: renamed `flavorDimensions` to `flavors`

**Review by brentvatne — APPROVED** [review]

File: packages/expo-cli/src/commands/run/android/runAndroid.ts [hunk 3]
```diff
@@ -101,9 +102,14 @@ async function resolveOptionsAsync(
...
-  const variant = options.variant.toLowerCase();
-  const apkDirectory = Android.getAPKDirectory(projectRoot);
-  const apkVariantDirectory = path.join(apkDirectory, variant);
+  // TODO: why would this be different? Can we get the different name?
+  const appName = 'app';
+
+  const apkDirectory = path.join(projectRoot, 'android', appName, 'build', 'outputs', 'apk');
+
+  // buildDeveloperTrust -> build, developer, trust (where developer, and trust are flavors).
+  const [buildType, ...flavors] = options.variant.split(/(?=[A-Z])/).map(v => v.toLowerCase());
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/run/android/runAndroid.ts:111]:
```suggestion
  // This won't work for non-standard flavor names like "myFlavor" would be treated as "my", "flavor".
  const [buildType, ...flavors] = options.variant.split(/(?=[A-Z])/).map(v => v.toLowerCase());
```

**Comment by Stukz** [comment]
hey @EvanBacon I was trying to use some flavors and launch them with `expo run:android --variant developDebug`

image

but it's giving me this issue

image

but on the other hand if I try to use `npx react-native` it only works doing `npx react-native run-android --variant developDebug --appIdSuffix dev`

not sure if `--appIdSuffix` it's a param that should be required in order to build

**Comment by EvanBacon** [comment]
Seems like it, maybe we could add a naive check to automatically find the `applicationIdSuffix`. Would you be interested in opening a PR to add this? ([source lives here](https://github.com/expo/expo/blob/main/packages/%40expo/cli/src/run/android/runAndroidAsync.ts)).

**Pull request merged**